HDDS-9918. [hsync] Remove block token from Ratis log once verified.

jojochuang opened the pull request:
## What changes were proposed in this pull request?
HDDS-9918. [hsync] Remove block token from Ratis log once verified

After [HDDS-9697](https://issues.apache.org/jira/browse/HDDS-9697), block token is not useful after leader DN verifies it, and does not need to be recorded in Ratis log.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9918

## How was this patch tested?

Stand up a docker compose cluster, verified that secure DataNode does not preserve encodedToken in Ratis log.
Unit tests passed

szetszwo commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java:
```diff
@@ -465,10 +467,15 @@ public TransactionContext startTransaction(RaftClientRequest request)
           ContainerProtos.Result.BLOCK_ALREADY_FINALIZED));
       return transactionContext;
     } else {
-      final Context context = new Context(proto, proto);
+      ContainerCommandRequestProto commitContainerCommandProto =
+          ContainerCommandRequestProto
+              .newBuilder(proto)
+              .clearEncodedToken()
+              .build();
+      final Context context = new Context(proto, commitContainerCommandProto);
       return builder
           .setStateMachineContext(context)
-          .setLogData(proto.toByteString())
+          .setLogData(commitContainerCommandProto.toByteString())
```
Let's combine both cases.  BTW, we should not call `build()` for the intermediate protos.
```java
+++ b/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java
@@ -422,39 +422,27 @@ public TransactionContext startTransaction(RaftClientRequest request)
       }
       return builder.build().setException(ioe);
     }
+
+    // once the token is verified, clear it from the proto
+    final ContainerCommandRequestProto.Builder protoBuilder = ContainerCommandRequestProto.newBuilder(proto)
+        .clearEncodedToken();
+
     if (proto.getCmdType() == Type.WriteChunk) {
       final WriteChunkRequestProto write = proto.getWriteChunk();
-      // create the log entry proto
-      final WriteChunkRequestProto commitWriteChunkProto =
-          WriteChunkRequestProto.newBuilder()
-              .setBlockID(write.getBlockID())
-              .setChunkData(write.getChunkData())
-              // skipping the data field as it is
-              // already set in statemachine data proto
-              .build();
-      ContainerCommandRequestProto commitContainerCommandProto =
-          ContainerCommandRequestProto
-              .newBuilder(proto)
-              .setPipelineID(gid.getUuid().toString())
-              .setWriteChunk(commitWriteChunkProto)
-              .setTraceID(proto.getTraceID())
-              .build();
       Preconditions.checkArgument(write.hasData());
       Preconditions.checkArgument(!write.getData().isEmpty());
-
-      final Context context = new Context(proto, commitContainerCommandProto);
-      return builder
-          .setStateMachineContext(context)
-          .setStateMachineData(write.getData())
-          .setLogData(commitContainerCommandProto.toByteString())
-          .build();
-    } else {
-      final Context context = new Context(proto, proto);
-      return builder
-          .setStateMachineContext(context)
-          .setLogData(proto.toByteString())
-          .build();
+      // skipping the data field as it is already set in statemachine data proto
+      final WriteChunkRequestProto.Builder commitWriteChunkProto = WriteChunkRequestProto.newBuilder(write)
+          .clearData();
+      protoBuilder.setWriteChunk(commitWriteChunkProto)
+          .setPipelineID(gid.getUuid().toString())
+          .setTraceID(proto.getTraceID());
+
+      builder.setStateMachineData(write.getData());
     }
+    return builder.setStateMachineContext(new Context(proto, protoBuilder.build()))
+        .setLogData(proto.toByteString())
+        .build();
 
   }
```

szetszwo reviewed (commented):
@jojochuang , thanks for working on this!  How about target this to the master branch instead of the hsync branch?

jojochuang commented:
Updated the PR based on Nicholas' suggestion and rebased on master.
I just realized this part of code is quite different from master and HDDS-7593 branch. I'll cherrypick the commit to HDDS-7593 branch once merged.

szetszwo commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java:
```diff
@@ -422,39 +422,26 @@ public TransactionContext startTransaction(RaftClientRequest request)
+      builder.setStateMachineData(write.getData());
     }
+    return builder.setStateMachineContext(new Context(proto, protoBuilder.build()))
+        .setLogData(proto.toByteString())
```
My fault:  the `setLogData` should use the new proto.
```java
    final ContainerCommandRequestProto containerCommandRequestProto = protoBuilder.build();
    return builder.setStateMachineContext(new Context(proto, containerCommandRequestProto))
        .setLogData(containerCommandRequestProto.toByteString())
        .build();
```

jojochuang commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java:
```diff
@@ -422,40 +422,27 @@ public TransactionContext startTransaction(RaftClientRequest request)
+          .clearData();
+      protoBuilder.setWriteChunk(commitWriteChunkProto)
+          .setPipelineID(gid.getUuid().toString())
+          .setTraceID(proto.getTraceID());
```
i don't think this line is needed but kept it as is. We need to have a test to verify that traces are propagated in the system correctly.

szetszwo reviewed (approved):
+1 the change looks good.
